### PR TITLE
Make intra links work inside trait impl block

### LINKED ISSUE #54172: rustdoc cannot resolve specific trait impl functions
```rust
trait Foo {
   /// This function generally does foo without side effects. 
   /// See [`SpecialFoo::foo()`] for an example.
   fn foo();
}

struct SpecialFoo {}

impl Foo for SpecialFoo {
    /// This is a good example of doing [`Foo::foo()`]
    fn foo() {}
}
```

There's no way to have the link `SpecialFoo:foo()` resolve in the rustdoc, as `SpecialFoo` does not contain this function. Unless there's a special syntax for referring to `SpecialFoo`'s `Foo` impl that would allow this.

### PR DESCRIPTION
**Pull request by ghost**
Closes #54172

### CODE REVIEW

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 1]
```diff
@@ -230,16 +232,45 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
...
                 ) => {
-                    let item = cx
-                        .tcx
-                        .inherent_impls(did)
-                        .iter()
-                        .flat_map(|imp| cx.tcx.associated_items(*imp).in_definition_order())
-                        .find(|item| item.ident.name == item_name);
+                    // We need item's parent to know if it's
+                    // trait impl or struct/enum/etc impl
+                    let mut item_parent = None;
```
Comment (GuillaumeGomez) [inline src/librustdoc/passes/collect_intra_doc_links.rs:237]: Why not:

```rust
let mut item_parent = item_opt
    .and_then(|item| self.cx.as_local_hir_id(item.def_id))
    .and_then(|item_hir| {
        let parent_hir = self.cx.tcx.hir().get_parent_item(item_hir);
        cx.tcx.hir().find(parent_hir);
    });
```

?

**Comment by GuillaumeGomez** [comment]
Thanks for the PR! Looks good globally, just a little code improvement to be done and it's good for me!

cc @rust-lang/rustdoc

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 1 above)
Comment (xliiv) [reply src/librustdoc/passes/collect_intra_doc_links.rs]: Thanks, done.

**Comment by xliiv** [comment]
@rust-lang/rustdoc are we merging ;)

**Comment by GuillaumeGomez** [comment]
Moving forward then.

@bors: r+

**Pull request merged**